Implements filter on ErrorMessage

Ziul opened the pull request:
fixes #29 and fixes #30

- support custom regex to filter
- support custom max value
- support to select the correct group of the regex

Signed-off-by: Luiz Oliveira <luiz.oliveira@ibm.com>

flaviostutz commented:
Thanks for the work!

Please change the attribute names according to the discussion at https://github.com/labbsr0x/servlet-monitor/issues/30.
The names would be "error-info-regex" and "error-info-maxsize"

I don't think we would need the "index" parameter for grouping regex as we won't be using groups in regex. The idea is for this regex to define the characters that would be accepted in output message (see JS code at the begining of https://github.com/labbsr0x/servlet-monitor/issues/30).

flaviostutz commented on src/main/java/br/com/labbs/monitor/filter/MetricsCollectorFilter.java:
```diff
@@ -57,7 +58,15 @@
     private static final String ERROR_MESSAGE_PARAM = "error-message";
     private static final String DEBUG = "debug";
     private static final String APPLICATION_VERSION = "application-version";
+    private static final String DEFAULT_FILTER_REGEX = "^([a-zA-z0-9 ]{0,120})";
```
Please change the attribute names according to the discussion at #30.
The names would be "error-info-regex" and "error-info-maxsize"

I don't think we would need the "index" parameter for grouping regex as we won't be using groups in regex. The idea is for this regex to define the characters that would be accepted in output message (see JS code at the begining of #30). Please use default regex from the description in #30.

flaviostutz commented on src/main/java/br/com/labbs/monitor/filter/MetricsCollectorFilter.java:
```diff
@@ -57,7 +58,15 @@
+    private static final String FILTER_MAX_SIZE_PARAM = "error-message-max-size";
     private final List<String> exclusions = new ArrayList<String>();
+    private int filter_group_index = 0;
+    private int filter_max_size = 120;
```
Too high default max size. Please check description at #30

flaviostutz commented on src/main/java/br/com/labbs/monitor/filter/MetricsCollectorFilter.java:
```diff
@@ -57,7 +58,15 @@
     private static final String DEBUG = "debug";
     private static final String APPLICATION_VERSION = "application-version";
+    private static final String DEFAULT_FILTER_REGEX = "^([a-zA-z0-9 ]{0,120})";
+    private static final String FILTER_REGEX_PARAM = "error-message-regex";
```
Change name

flaviostutz commented on src/main/java/br/com/labbs/monitor/filter/MetricsCollectorFilter.java:
```diff
@@ -57,7 +58,15 @@
     private static final String APPLICATION_VERSION = "application-version";
+    private static final String DEFAULT_FILTER_REGEX = "^([a-zA-z0-9 ]{0,120})";
+    private static final String FILTER_REGEX_PARAM = "error-message-regex";
+    private static final String FILTER_GROUP_INDEX_PARAM = "error-message-regex-index";
```
I think this is unnecessary because we didn't discuss about using regex groupings.

flaviostutz commented on src/main/java/br/com/labbs/monitor/filter/MetricsCollectorFilter.java:
```diff
@@ -57,7 +58,15 @@
+    private static final String DEFAULT_FILTER_REGEX = "^([a-zA-z0-9 ]{0,120})";
+    private static final String FILTER_REGEX_PARAM = "error-message-regex";
+    private static final String FILTER_GROUP_INDEX_PARAM = "error-message-regex-index";
+    private static final String FILTER_MAX_SIZE_PARAM = "error-message-max-size";
```
Please use "error-info-maxsize" at in #30

flaviostutz commented on src/main/java/br/com/labbs/monitor/filter/MetricsCollectorFilter.java:
```diff
@@ -57,7 +58,15 @@
+    private static final String FILTER_GROUP_INDEX_PARAM = "error-message-regex-index";
+    private static final String FILTER_MAX_SIZE_PARAM = "error-message-max-size";
     private final List<String> exclusions = new ArrayList<String>();
+    private int filter_group_index = 0;
```
we won't use according to previous discussion in #30

flaviostutz commented on src/main/java/br/com/labbs/monitor/filter/MetricsCollectorFilter.java:
```diff
@@ -57,7 +58,15 @@
     private final List<String> exclusions = new ArrayList<String>();
+    private int filter_group_index = 0;
+    private int filter_max_size = 120;
+    private String filter_regex = "";
```
We would have a default regex according to #30

Ziul commented:
Updated the attribute names according to #30.

The index parameter was for select any group of the regex. Now that I updated the PR to apply the regex as a whitelist it's pointless, so I removed the index. Also reduced the default max size according to #29.

flaviostutz commented on README.md:
```diff
@@ -174,6 +174,10 @@ Your java code should look like this:
 
+###### Improving error messages
+
+It is possible to filter the error message to avoid long messages or personal info exposed in the metrics. To do it, three params may be used: `error-info-regex`, `error-info-regex-index`, and `error-info-max-size`. The first will set the regex to apply in the message, with `^([a-zA-z0-9 ]{0,120})` as the default value. The second, `error-info-regex-index`, selects the group of the regex's result ...
```
I think we have to remove the "error-info-regex-index" mention here as it is not used anymore.

flaviostutz commented on src/main/java/br/com/labbs/monitor/filter/MetricsCollectorFilter.java:
```diff
@@ -209,20 +222,36 @@ private boolean isErrorStatus(int status) {
      * If error message is null, sets the string to empty string.
+     * If a regex is defined, use it to filter message
+     * 
+     * Default regex: ^([a-zA-z0-9 ]{0,120})
```
Documentation is wrong here.

flaviostutz replied:
Default regex is wrong here in documentation.

flaviostutz commented on src/main/java/br/com/labbs/monitor/filter/MetricsCollectorFilter.java:
```diff
@@ -209,20 +222,36 @@ private boolean isErrorStatus(int status) {
+                result = result.substring(0, filter_max_size);
+            }
+        } catch (Exception e) {
+            // avoid invalid regex or invalid matcher group index
```
Log here the exception with "warning" for the app developer to know that something is wrong.

flaviostutz commented on src/main/java/br/com/labbs/monitor/filter/MetricsCollectorFilter.java:
```diff
@@ -57,7 +58,14 @@
     private static final String ERROR_MESSAGE_PARAM = "error-message";
     private static final String DEBUG = "debug";
     private static final String APPLICATION_VERSION = "application-version";
+    private static final String DEFAULT_FILTER_REGEX = "[^A-zÀ-ú .,]+";
```
Check if " " is accepted. It would be better to use "\s" for representing spaces in regex.

flaviostutz reviewed (changes requested):
Thanks for the changes... we are almost there... see review comments.

flaviostutz reviewed (approved):
Thanks, man!
